### ISSUE TITLE
[Bug]: Incompatible with Svelte 5 (pre-release)

### ISSUE THREAD

**Issue by anowell** [SauravKanchan/svelte-chartjs#158]
### Would you like to work on a fix?

- [ ] Check this if you would like to implement a PR, we are more than happy to help you go through the process.

### Current and expected behavior

Importing a chartjs component (e.g. Bar) with Svelte 5 (svelte@5.0.0-next.115) yields this error message:

> Your application, or one of its dependencies, imported from 'svelte/internal', which was a private module used by Svelte 4 components that no longer exists in Svelte 5. It is not intended to be public API. If you're a library author and you used 'svelte/internal' deliberately, please raise an issue on https://github.com/sveltejs/svelte/issues detailing your use case.

I haven't looked any further than simply confirming the import in this library:

https://github.com/SauravKanchan/svelte-chartjs/blob/4333128d3f47b7e8c9a7a77177ad8592c2e22999/src/utils/svelte.ts#L1-L6

### Reproduction

[did not create shareable repro - merely confirmed svelte/internal is import]

### chart.js version

4.4.2

### svelte-chartjs version

 3.1.5

**Comment by shahedsalehi** [comment]
I just updated my project to svelte svelte@5.0.0-next.123 and I encountered the same issue.

**Comment by andupotorac** [comment]
Same issue here, this doesn't work for Svelte 5.

**Comment by avi12** [comment]
Svelte 5 was released, this library doesn't work
For the time being, I'll use Svelte 4, but if this library doesn't support 5, I'll ditch it at some point

**Comment by PierrickP** [comment]
@anowell Can you propose a proper PR here (without `console.log`) ? :pray:

**Comment by MarcGodard** [comment]
This project isn't very active. If someone forks and fix, please let this thread know. If its not fixed by next month, I will do it.

**Comment by avi12** [comment]
Following the repository's source code, I managed to replicate the behavior in Svelte 5
TL;DR
```html
<script lang="ts">
  import { BarController, BarElement, CategoryScale, Chart, LinearScale, TimeScale, Tooltip } from "chart.js";

  import "chartjs-adapter-date-fns";

  let elCanvas: HTMLCanvasElement;

  Chart.register(Tooltip, BarElement, BarController, CategoryScale, LinearScale, TimeScale);

  let chart: Chart;

  onMount(() => {
    chart = new Chart(elCanvas, {
      type: "bar",
      data,
      options: {
        scales: {
          y: {
            grace: "5%"
          },
          x: {
            ticks: {
              callback(i) {
                // ...
              },
              color
            }
          }
        },
        plugins: {
          legend: {
            display: false
          },
          tooltip: {
            callbacks: {
              title(tooltipItem) {
                // ...
              }
            }
          }
        },
        animation: false
      }
    });
  });

  const themeInUse = $derived("light");
  const color = $derived(themeInUse === "dark" ? "#fffFFF" : "#1B1B1B");

  $effect(() => {
    if (chart) {
      chart.data = data;
      chart.update();
    }
  });

  $effect(() => {
    if (chart) {
      chart.config.options.scales.x.ticks.color = color;
      chart.config.options.scales.y.ticks.color = color;
      chart.update();
    }
  });

  // https://www.chartjs.org/docs/latest/charts/bar.html
  const data = $derived({
    datasets: [
      {
        data: (() => {
          // ...
          }, []);
        })(),
        backgroundColor: "#1BE7FF"
      }
    ]
  });
</script>

<canvas bind:this={elCanvas}></canvas>
```

**Comment by fev4** [comment]
took inspiration from @avi12 (thank you btw) and here's a complete working component for <Line/> working in Svelte 5 that you can just drop in as a replacement

my hope is that you can use this `Line` component as a guide to create every other component you might need

in any case, here it is:

```
<script lang="ts">
  import {
    Chart,
    Tooltip,
    type ChartData,
    type ChartOptions,
  } from 'chart.js';
  import type { HTMLCanvasAttributes } from 'svelte/elements';

  import 'chart.js/auto';
  import 'chartjs-adapter-date-fns';

  interface Props extends HTMLCanvasAttributes {
    data: ChartData<'line', number[], string>;
    options: ChartOptions<'line'>;
  }

  const { data, options, ...rest }: Props = $props();

  Chart.register(Tooltip);

  let canvasElem: HTMLCanvasElement;
  let chart: Chart;

  $effect(() => {
    chart = new Chart(canvasElem, {
      type: 'line',
      data,
      options,
    });

    return () => {
      chart.destroy();
    };
  });

  $effect(() => {
    if (chart) {
      chart.data = data;
      chart.update();
    }
  });
</script>

<canvas bind:this={canvasElem} {...rest}></canvas>
``` 

note a couple of things:
- uses Chart.register() to only register the things I personally need, you might need others
- doesn't use `onMount` instead only `$effect` calls
- it returns on one of the `$effect` to clean up the chart when doing internal transitions, in order to avoid errors like `Uncaught Error: Canvas is already in use. Chart with ID '15' must be destroyed before the canvas with ID '' can be reused.`
- uses typescript to extend the `HTMLCanvasAttribute` so that users of the `Line` component can pass props like `height`, and so on
- it also uses specific types to the `Line` chart from `chart.js` types
- I'm importing some extensions like `chart.js/auto` and `chartjs-adapter-date-fns`, you might not need those
- it's a very simple implementation but that's mostly why I like it

IMHO `svelte-chartjs` is not needed anymore as with the simplicity and extensibility Svelte 5 offers, a simple `Line` component can be created as seen above, `svelte-chartjs` can still have a place but I believe it would need to offer some other differentiation

**Comment by avi12** [comment]
@fev4 According to my testing, you must put the `new Chart` within an `onMount`, or otherwise it will complain that more than a single chart is being initialized

**Comment by dr-starchildren** [comment]
I'm going through the similar transition to Svelte 5 and having this break for me. I based my modifications on @fev4's design (Thanks for sharing!). I had a more static chart, but changing data, so I broke out my data into dataPoints and labels.
```javascript
<script lang="ts">
	import ChartDataLabels from 'chartjs-plugin-datalabels';
	import { Chart, Title, Tooltip, Legend, ArcElement, CategoryScale } from 'chart.js';
	import 'chart.js/auto';

	const { dataPoints, labels, ...rest } = $props();
	Chart.register(Title, Tooltip, Legend, ArcElement, CategoryScale, ChartDataLabels);

	const data = $derived({
		labels: labels,
		datasets: [
			{
				data: dataPoints,
				backgroundColor: [
					'#EE714F', '#2894B5', '#8F3C8F', '#445B62', '#FDD50E', '#DB8872', 
					'#5A99AD', '#804F80','#6F848B', '#FCE366'
				]
			}
		]
	});
	let canvasElem: HTMLCanvasElement;
	let chart: Chart;

	$effect(() => {
		chart = new Chart(canvasElem, {
			type: 'pie',
			data,
			options: { responsive: true }
		});

		return () => {
			chart.destroy();
		};
	});

	$effect(() => {
		if (chart) {
			chart.data = data;
			chart.update();
		}
	});

	Chart.defaults.set('plugins.datalabels', {
		color: '#FFFFFF',
		formatter: function (value, context) {
			return context.chart.data.labels[context.dataIndex];
		},
		font: { size: '24' },
		color: 'white',
		display: function (context) {
			var dataset = context.dataset;
			var count = dataset.data.length;
			var value = dataset.data[context.dataIndex];
			return value > 10;
		},
		padding: 6
	});
	Chart.defaults.set('plugins.legend', { display: false });
</script>

<div>
	<canvas bind:this={canvasElem} {...rest}></canvas>
</div>

<style>
	div {
		width: var(--width, 300px);
		padding-bottom: 16px;
	}
</style>
```

I use it like this:
```javascript
<Pie dataPoints={dataPointsFreq} {labels} />
```

**Comment by LupusAI** [comment]
Hi everyone, since this repo doesn’t seem to be actively maintained, I’ve created a Svelte 5 compatible fork of the library: 

https://github.com/LupusAI/svelte5-chartjs.git 

Feel free to check it out and contribute!

**Comment by Zalcherei** [comment]
For those looking for minimum setup for Svelte 5:

chart.svelte component:
```html
<script lang="ts">
  import { onMount } from 'svelte';
  import Chart from 'chart.js/auto';
  
  let { type, data, options } = $props();
  let canvas: HTMLCanvasElement;
  
  onMount(() => {
    if (canvas) {
      new Chart(canvas, {
        type: type,
        data: data,
        options: options
      });
    }
  });
</script>

<div class="relative h-fit w-full">
  <canvas bind:this={canvas}></canvas>
</div>
```

Usage:
```html
<script>
  import Chart from '$lib/chartjs/chart.svelte';
  
  const chartData = {
    labels: ['Red', 'Blue', 'Yellow', 'Green', 'Purple', 'Orange'],
    datasets: [
    	{
          label: 'Sample Data',
            data: [12, 19, 3, 5, 2, 3],
            backgroundColor: [
              'rgba(255, 99, 132, 0.2)',
              'rgba(54, 162, 235, 0.2)',
              'rgba(255, 206, 86, 0.2)',
              'rgba(75, 192, 192, 0.2)',
              'rgba(153, 102, 255, 0.2)',
              'rgba(255, 159, 64, 0.2)'
            ],
            borderColor: [
              'rgba(255, 99, 132, 1)',
              'rgba(54, 162, 235, 1)',
              'rgba(255, 206, 86, 1)',
              'rgba(75, 192, 192, 1)',
              'rgba(153, 102, 255, 1)',
              'rgba(255, 159, 64, 1)'
            ],
          borderWidth: 1
    	}
    ]
  };
  
  const chartOptions = {
    responsive: true,
    scales: {
      y: {
        beginAtZero: true
      }
    }
  };
</script>

<Chart type="bar" data={chartData} options={chartOptions} />
```

**Comment by SauravKanchan** [comment]
This has been resolved by merging #170, which is a complete Svelte 5 migration. The library no longer imports from `svelte/internal` and now uses Svelte 5 runes throughout.

Please update to the latest version. Thanks for reporting!
